sklearn/linear_model/tests/test_logistic.py

```diff
@@ -1379,6 +1394,247 @@ def test_warm_start_converge_LR():
+def test_elastic_net_l1_l2_equivalence(C):
+    # Make sure elasticnet is equivalent to l1 when l1_ratio=1 and l2 when
+    # l1_ratio=0.
+    X, y = make_classification(random_state=0)
+
+    lr = LogisticRegression(penalty='elasticnet', C=C, l1_ratio=0,
+                            solver='saga', random_state=0)
+    lr.fit(X, y)
+    elastic_net_l1_zero_coeffs = lr.coef_
+
+    lr = LogisticRegression(penalty='l2', C=C, solver='saga', random_state=0)
+    lr.fit(X, y)
+    l2_coeffs = lr.coef_
+
+    assert_array_almost_equal(elastic_net_l1_zero_coeffs, l2_coeffs)
+
+    lr = LogisticRegression(penalty='elasticnet', C=C, l1_ratio=1,
+                            solver='saga', random_state=0)
+    lr.fit(X, y)
+    elastic_net_l1_one_coeffs = lr.coef_
+
+    lr = LogisticRegression(penalty='l1', C=C, solver='saga', random_state=0)
+    lr.fit(X, y)
+    l1_coeffs = lr.coef_
+
+    assert_array_almost_equal(elastic_net_l1_one_coeffs, l1_coeffs)
+
+
+@pytest.mark.parametrize('C', [.001, 1, 100, 10000000])
+def test_elastic_net_vs_l1_l2(C):
+    # Make sure that elastic net with grid search on l1_ratio gives same or
+    # better results than just l1 or just l2.
+
+    X, y = make_classification(500, random_state=0)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
+
+    param_grid = {'l1_ratio': np.linspace(0, 1, 10)}
+
+    enet_clf = LogisticRegression(penalty='elasticnet', C=C, solver='saga',
+                                  random_state=0)
+    gs = GridSearchCV(enet_clf, param_grid, cv=5, iid=False, refit=True)
+
+    l1_clf = LogisticRegression(penalty='l1', C=C, solver='saga',
+                                random_state=0)
+    l2_clf = LogisticRegression(penalty='l2', C=C, solver='saga',
+                                random_state=0)
+
+    for clf in (gs, l1_clf, l2_clf):
+        clf.fit(X_train, y_train)
+
+    assert gs.score(X_test, y_test) >= l1_clf.score(X_test, y_test)
+    assert gs.score(X_test, y_test) >= l2_clf.score(X_test, y_test)
+
+
+@pytest.mark.parametrize('C', np.logspace(-3, 2, 4))
+@pytest.mark.parametrize('l1_ratio', [.5])
+def test_LogisticRegression_elastic_net_objective(C, l1_ratio):
+    # train a logistic regression with l2 (a) and elasticnet (b) penalties,
+    # and compute the elasticnet objective. That of a sould be greater than
+    # that of b.
+    X, y = make_classification(n_samples=10000, n_classes=2, n_features=20,
+                               n_informative=10, n_redundant=0,
+                               n_repeated=0, random_state=0)
+    X = scale(X)
+
+    lr_enet = LogisticRegression(penalty='elasticnet', solver='saga',
+                                 random_state=0, C=C, l1_ratio=l1_ratio)
+    lr_l2 = LogisticRegression(penalty='l2', solver='saga', random_state=0,
+                               C=C)
+    lr_enet.fit(X, y)
+    lr_l2.fit(X, y)
+
+    def enet_objective(lr):
+        obj = C * log_loss(y, lr.predict(X))
```

Comment:
needs to be ``predict_proba``

Reply:
oops thanks!
Still failing though :/ , same for SGDClassifier

Reply:
it should be lr.decision_function